languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+# Roles 2
+from typing import Any, List, Set
+from dataclasses import dataclass
+
+from oso import Variable
+
+from sqlalchemy.types import Integer, String
+from sqlalchemy.schema import Column, ForeignKey
+from sqlalchemy import inspect
+
+
+# Starting with something like this, tbd
+
+
+@dataclass
+class Relationship:
+    child_python_class: Any
+    child_type: str
+    child_table: str
+    parent_python_class: Any
+    parent_type: str
+    parent_table: str
+    parent_selector: Any
+    parent_field: str
+
+
+@dataclass
+class Permission:
+    python_class: Any
+    name: str
+
+
+@dataclass
+class Role:
+    name: str
+    python_class: Any
+    permissions: List[Permission]
+    implied_roles: List[str]
+
+
+@dataclass
+class Resource:
+    name: str
+    python_class: Any
+    actions: Set[str]
+    roles: Set[str]
+
+
+def ensure_configured(func):
+    def wrapper(self, *args, **kwargs):
+        if not self.configured:
+            self._configure()
+        func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class OsoRoles:
+    def __init__(self, oso, sqlalchemy_base, user_model):
+        # @Q: Is this where I should create the models?
```

Comment:
We also might want to think about how migrations would work. SQLAlchemy doesn't really have a standard migration tool, although I think a lot of people use Alembic.

Not for this first release... but we probably need to figure it out before the next one.